library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java

```diff
@@ -0,0 +1,264 @@
+		return this;
+	}
+
+	/**
+	 * Sets the default identifier of this registry.
+	 * <p>
+	 * Should a nonexistent entry be referenced in some way, the registry will instead reference the entry identified
+	 * by this instead.
+	 * <p>
+	 * By default, this is {@code null} - the registry will simply return {@code null} when a nonexistent entry is referenced.
+	 *
+	 * @param defaultId the new default identifier
+	 * @return this builder
+	 */
+	@Contract("_ -> this")
+	public @NotNull QuiltRegistryBuilder<T> withDefaultId(@Nullable Identifier defaultId) {
+		this.defaultId = defaultId;
+		return this;
+	}
+
+	/**
+	 * Sets the synchronization behavior of this registry.
+	 * <p>
+	 * By default, this is {@link SyncBehavior#SKIPPED}.
+	 *
+	 * @param syncBehavior the new synchronization behavior
+	 * @return this builder
+	 */
+	@Contract("_ -> this")
+	public @NotNull QuiltRegistryBuilder<T> withSyncBehavior(@NotNull QuiltRegistryBuilder.SyncBehavior syncBehavior) {
+		this.syncBehavior = syncBehavior;
+		return this;
+	}
+
+	/**
+	 * Sets the registry to <em>not</em> be synchronized at all.
+	 *
+	 * @return this builder
+	 * @see #withSyncBehavior(SyncBehavior)
+	 * @see SyncBehavior#SKIPPED
+	 */
+	@Contract("-> this")
+	public @NotNull QuiltRegistryBuilder<T> withSyncSkipped() {
+		return this.withSyncBehavior(SyncBehavior.SKIPPED);
+	}
+
+	/**
+	 * Sets the registry to be synchronized, and to be required - clients who do not have this registry on their side
+	 * <em>will</em> be kicked.
+	 *
+	 * @return this builder
+	 * @see #withSyncBehavior(SyncBehavior)
+	 * @see SyncBehavior#REQUIRED
+	 */
+	@Contract("-> this")
+	public @NotNull QuiltRegistryBuilder<T> withSyncRequired() {
+		return this.withSyncBehavior(SyncBehavior.REQUIRED);
+	}
+
+	/**
+	 * Sets the registry to be synchronized, and to be optional - clients who do not have this registry on their side
+	 * <em>will not</em> be kicked.
+	 *
+	 * @return this builder.
+	 * @see #withSyncBehavior(SyncBehavior)
+	 * @see SyncBehavior#OPTIONAL
+	 */
+	@Contract("-> this")
+	public @NotNull QuiltRegistryBuilder<T> withSyncOptional() {
+		return this.withSyncBehavior(SyncBehavior.OPTIONAL);
+	}
+
+	/**
+	 * Builds the {@code Registry} instance.
+	 *
+	 * @return the newly constructed registry
+	 */
+	@SuppressWarnings("unchecked")
+	@Contract("-> new")
+	public @NotNull Registry<T> build() {
+		SimpleRegistry<T> registry;
+		if (this.defaultId == null) {
+			registry = new SimpleRegistry<>(this.key, this.lifecycle, this.useIntrusiveHolders);
+		} else {
+			// this takes the identifier as a string, to guarantee that it's unique
+			registry = new DefaultMappedRegistry<>(this.defaultId.toString(), this.key, this.lifecycle, this.useIntrusiveHolders);
+		}
+
+		Registry.register((Registry<Registry<Object>>) Registries.ROOT, this.key.getValue(), (Registry<Object>) registry);
+
+		if (this.syncBehavior == SyncBehavior.REQUIRED || this.syncBehavior == SyncBehavior.OPTIONAL) {
```

Comment:
```suggestion
		if (this.syncBehavior != SyncBehavior.SKIPPED) {
```
seems like this is more to the point